File: rca_core/error_utils.py
"""Error normalization and retry utilities.

Inspired by DeepSeek Harness's pi-ai error handling patterns:
- Unified error normalization across different LLM providers
- Smart retry with jitter and Retry-After header support
- Consistent error formatting for user display

This module provides:
- NormalizedError: structured error representation
- retry_with_backoff(): exponential backoff with jitter
- format_provider_error(): human-readable error messages
"""

from __future__ import annotations

import datetime
import json
import random
import time
from dataclasses import dataclass
from email.utils import parsedate_to_datetime
from typing import Callable, TypeVar

T = TypeVar("T")

# Status codes worth retrying (transient upstream errors).
# 401/403/400 are authentication/client errors - retrying is futile.
RETRYABLE_STATUS = {408, 425, 429, 500, 502, 503, 504}

# Maximum error body length to store.
MAX_ERROR_BODY_CHARS = 2000

# REVIEW-2026-09-20 #108: floor for every computed retry delay. A provider
# legitimately answers a 429 with ``Retry-After: 0`` (or ``Retry-After-Ms: 0``,
# or an HTTP-date that has already passed), which used to be honoured
# literally: the retry loop then hammered the endpoint back-to-back with zero
# spacing, which is exactly what turns a rate limit into a longer BAN. One
# second is also the smallest delay that survives the ``Retry-After`` round
# trip of most gateways, and a caller that genuinely wants no wait can still
# pass ``max_delay=0`` (the clamp is applied BELOW the ceiling, so a ceiling
# under the floor wins — tests rely on that).
MIN_RETRY_DELAY_SECONDS = 1.0

@dataclass
class NormalizedError:
    """Structured error representation across different provider error formats.

    Mirrors the pattern from DSH's pi-ai error-body.js normalization.
    """

    # HTTP status code (None for network errors)
    status: int | None = None

    # Decoded error body string
    body: str = ""

    # Original error message
    message: str = ""

    # Whether the message already contains the body
    message_carries_body: bool = False

    # Provider-specific error code
    error_code: str | None = None

    # Retry suggestion (None, "retry", "do_not_retry")
    retry_hint: str | None = None

    @property
    def is_retryable(self) -> bool:
        """Check if this error should be retried."""
        if self.status is None:
            return True
        return self.status in RETRYABLE_STATUS

    @property
    def display_message(self) -> str:
        """Get the best message for user display."""
        if self.message_carries_body or not self.body:
            base = self.message
        else:
            prefix = f"HTTP {self.status}" if self.status else "Network error"
            base = f"{prefix}: {self.body}"
        if self.error_code:
            base = f"[{self.error_code}] {base}"
        return base


def normalize_http_error(
    status: int | None,
    body_bytes: bytes | None,
    message: str,
) -> NormalizedError:
    """Normalize an HTTP error into a structured form."""
    body = _decode_body(body_bytes)
    stored = body[:MAX_ERROR_BODY_CHARS]
    # REVIEW-2026-11-07 (low): check the STORED (truncated) form, not the
    # full decoded body. Before, a message that embedded the body as the
    # caller had truncated it (or as the retry suffix re-emitted it) failed
    # the `message.find(full_body)` test, display_message then preferred
    # the raw prefix path, and the message's own context was lost. A
    # message containing the full body also contains its truncated prefix,
    # so testing `stored` covers both cases.
    message_carries_body = (not stored) or (stored in message)
    error_code = _extract_error_code(body_bytes, body)
    retry_hint = _get_retry_hint(status, error_code)

    return NormalizedError(
        status=status,
        body=stored,
        message=message,
        message_carries_body=message_carries_body,
        error_code=error_code,
        retry_hint=retry_hint,
    )


def _decode_body(body_bytes: bytes | None) -> str:
    """Best-effort decode of error body bytes."""
    if not body_bytes:
        return ""
    for encoding in ("utf-8", "latin-1"):
        try:
            return body_bytes.decode(encoding, errors="replace")[:MAX_ERROR_BODY_CHARS]
        except Exception:
            continue
    return ""


def _extract_error_code(body_bytes: bytes | None, body: str) -> str | None:
    """Try to extract a machine-readable error code from the body.

    REVIEW-2026-09-20 #110: this used to re-decode ``body_bytes`` and parse
    the FULL payload while ``_decode_body`` had already truncated the body it
    stores/returns at ``MAX_ERROR_BODY_CHARS`` — so the ``[CODE]`` prefix the
    user sees could come from text that is no longer displayed (and a huge
    hostile body cost a second, unbounded parse). ``body`` is now the single
    source of truth: what gets parsed is exactly what is stored, which is also
    what makes the code consistent with the ``display_message`` the operator
    can actually correlate against a provider status page. A body truncated
    mid-JSON simply yields no code, like any other non-JSON error body.
    ``body_bytes`` stays in the signature for the existing callers.
    """
    if not body:
        return None
    try:
        data = json.loads(body)
        if isinstance(data, dict):
            for key in ("error_code", "code", "type", "error.type"):
                if key in data:
                    return str(data[key])
            error = data.get("error", {})
            if isinstance(error, dict):
                for key in ("error_code", "code", "type"):
                    if key in error:
                        return str(error[key])
    except Exception:
        pass
    return None


def _get_retry_hint(status: int | None, error_code: str | None) -> str | None:
    """Determine if and how to retry based on status and error code."""
    if status is None:
        return "retry"
    if status == 429:
        return "retry_with_delay"
    if status in (401, 403, 400):
        return "do_not_retry"
    if 500 <= status < 600:
        return "retry"
    if status in RETRYABLE_STATUS:
        return "retry"
    return None


def format_provider_error(
    normalized: NormalizedError,
    prefix: str | None = None,
) -> str:
    """Format a normalized error for user display."""
    msg = normalized.display_message
    if prefix and normalized.status:
        return f"{prefix} ({normalized.status}): {msg}"
    if normalized.status:
        return f"HTTP {normalized.status}: {msg}"
    return msg

def get_retry_delay(
    status: int | None,
    headers: dict | None = None,
    attempt: int = 0,
    initial_delay: float = 0.8,
    backoff_factor: float = 1.6,
    max_delay: float = 60.0,
) -> float:
    """Calculate retry delay with exponential backoff and jitter.

    Respects Retry-After headers when present, with smart fallback.
    Mirrors DSH's provider-retry.js getRetryDelayMs() logic.
    """
    delay: float | None = None

    # 1. Check Retry-After-MS header (milliseconds)
    if headers:
        retry_after_ms = headers.get("retry-after-ms") or headers.get("Retry-After-Ms")
        if retry_after_ms:
            try:
                delay = float(retry_after_ms) / 1000.0
            except (ValueError, TypeError):
                pass

        # 2. Check Retry-After header (seconds or HTTP date)
        if delay is None:
            retry_after = headers.get("retry-after") or headers.get("Retry-After")
            if retry_after:
                try:
                    delay = float(retry_after)
                except ValueError:
                    try:
                        dt = parsedate_to_datetime(retry_after)
                        delay = max(0.0, (dt - datetime.datetime.now(datetime.timezone.utc)).total_seconds())
                    except Exception:
                        pass

    # 3. Fall back to exponential backoff with jitter
    if delay is None:
        base_delay = initial_delay * (backoff_factor ** attempt)
        # Add jitter: DSH pattern: multiply by (1 - random * 0.25)
        jitter = 1.0 - random.random() * 0.25
        delay = base_delay * jitter

    # REVIEW-2026-09-20 #108: cap first, then lift to the 1s floor. A
    # ``Retry-After: 0`` (or an already-elapsed HTTP date) is honoured as
    # "no information" rather than as "hammer me again immediately"; a caller
    # that caps below the floor (``max_delay=0``, i.e. the tests that must not
    # sleep) still gets its ceiling because the floor is clamped to it.
    delay = min(delay, max_delay)
    return max(delay, min(MIN_RETRY_DELAY_SECONDS, max_delay))


def retry_with_backoff(
    func: Callable[[], T],
    *,
    max_retries: int = 3,
    initial_delay: float = 0.8,
    backoff_factor: float = 1.6,
    max_delay: float = 60.0,
    retryable: Callable[[T], bool] | None = None,
    on_retry: Callable[[int, float, Exception | None], None] | None = None,
) -> T:
    """Execute a function with exponential backoff retry.

    Args:
        func: The function to execute.
        max_retries: Maximum number of retry attempts.
        initial_delay: Initial delay in seconds.
        backoff_factor: Exponential backoff multiplier.
        max_delay: Maximum delay cap in seconds.
        retryable: Optional predicate to determine if result warrants retry.
            ``None`` means every result is acceptable (return immediately).
        on_retry: Optional callback called before each retry.

    Returns:
        The function's return value. When ``retryable`` keeps returning
        True and the attempts are exhausted without an exception, the LAST
        observed result is returned.

    Raises:
        The last exception if all retries fail and an exception occurred.

    Sprint B (REVIEW-2026-09-04): implemented the documented predicate
    semantics. The previous code returned ``result`` in BOTH branches of
    the check (``if retryable(result): return result`` / ``return result``),
    so a result the predicate deemed retry-worthy never actually triggered
    another attempt — the retryable parameter was dead logic. Now:
    ``retryable(result)`` False -> return the result immediately (final);
    True -> keep retrying until the attempts are exhausted.

    REVIEW-2026-09-20 #109: the loop used to sleep after EVERY iteration that
    did not return, including the LAST attempt when the failure came from the
    ``retryable`` predicate rather than an exception (the exception path
    already ``break``-ed out before the sleep). Callers therefore paid one
    more back-off wait for a retry that was never going to happen — with the
    defaults a 3-retry loop that kept answering "retryable" slept 4 times
    instead of 3. The final attempt now breaks out before the delay is
    computed, so ``on_retry`` and ``sleep`` run exactly once per ACTUAL retry.
    """
    last_exc: Exception | None = None
    last_result: T | None = None

    for attempt in range(max_retries + 1):
        try:
            result = func()
            # Check if result is acceptable (retryable=None means all results
            # are acceptable). A non-retryable result is final, NOT an error:
            # return it directly. A retryable result means "keep trying".
            if retryable is None or not retryable(result):
                return result
            last_result = result
        except Exception as exc:
            last_exc = exc
            if attempt >= max_retries:
                break

        if attempt >= max_retries:
            # Last attempt and the result was retryable: nothing follows, so
            # do not sleep once more before returning it.
            break

        # Calculate delay for next retry
        delay = initial_delay * (backoff_factor ** attempt)
        jitter = 1.0 - random.random() * 0.25
        delay *= jitter
        delay = min(delay, max_delay)

        if on_retry:
            on_retry(attempt + 1, delay, last_exc)

        time.sleep(delay)

    # All retries exhausted
    if last_exc is not None:
        raise last_exc
    # No exception occurred but every result was deemed retryable.
    # Return the last observed result (None only if func never returned,
    # which cannot happen without raising).
    return last_result


def retry_http_request(
    request_func: Callable[[], tuple[int | None, bytes | None, bytes | None]],
    *,
    max_retries: int = 3,
    initial_delay: float = 0.8,
    backoff_factor: float = 1.6,
    max_delay: float = 60.0,
    get_headers: Callable[[], dict | None] | None = None,
) -> tuple[int | None, bytes | None, bytes | None]:
    """Retry an HTTP request with smart backoff.

    Specifically designed for LLM API calls. Handles Retry-After headers.
    """
    for attempt in range(max_retries + 1):
        status, body, err_body = request_func()

        # Success
        if err_body is None or len(err_body) == 0:
            if status is None or status < 400:
                return status, body, err_body

        # Check if retryable
        if status is not None and status not in RETRYABLE_STATUS:
            return status, body, err_body

        # Network error - only retry once more
        if status is None and attempt >= 1:
            return status, body, err_body

        if attempt >= max_retries:
            return status, body, err_body

        # Get headers for Retry-After
        headers = get_headers() if get_headers else None

        # Calculate delay
        delay = get_retry_delay(
            status=status,
            headers=headers,
            attempt=attempt,
            initial_delay=initial_delay,
            backoff_factor=backoff_factor,
            max_delay=max_delay,
        )

        # Add retry annotation to error body
        suffix = f"[retry {attempt + 1}/{max_retries + 1} after {delay:.1f}s]"
        if err_body:
            err_body = err_body + suffix.encode("utf-8") if isinstance(err_body, bytes) else (err_body + suffix).encode("utf-8")
        else:
            err_body = suffix.encode("utf-8")

        time.sleep(delay)

    return status, body, err_body

File: js/error-utils.js
// error-utils.js - Error normalization and retry utilities for frontend.
// Inspired by DeepSeek Harness's pi-ai error handling patterns:
// - Unified error normalization across different LLM providers
// - Smart retry with jitter and Retry-After header support
// - Consistent error formatting for user display

'use strict';

// ---------------------------------------------------------------------------
// Error Normalization
// ---------------------------------------------------------------------------

/**
 * @typedef {Object} NormalizedError
 * @property {number|null} status - HTTP status code (null for network errors)
 * @property {string} body - Decoded error body string
 * @property {string} message - Original error message
 * @property {boolean} messageCarriesBody - Whether message already contains body
 * @property {string|null} errorCode - Provider-specific error code
 * @property {string|null} retryHint - "retry" | "retry_with_delay" | "do_not_retry" | null
 */

/** Status codes worth retrying (transient upstream errors). */
const RETRYABLE_STATUS = new Set([408, 425, 429, 500, 502, 503, 504]);

/** Maximum error body length to store. */
const MAX_ERROR_BODY_CHARS = 2000;

/**
 * REVIEW-2026-09-20 #108 (mirror of rca_core/error_utils.py
 * MIN_RETRY_DELAY_SECONDS): floor for every computed retry delay. A provider
 * legitimately answers a 429 with ``Retry-After: 0`` (or ``Retry-After-Ms: 0``,
 * or an HTTP-date that has already passed), which used to be honoured
 * literally: the retry loop then hammered the endpoint back-to-back with zero
 * spacing, which is exactly what turns a rate limit into a longer BAN.
 * The floor is applied BELOW the ceiling, so a caller that caps under it
 * (``maxDelay = 0``, i.e. the tests that must not sleep) still gets its
 * ceiling — see getRetryDelay().
 */
const MIN_RETRY_DELAY_SECONDS = 1.0;

/**
 * Normalize an HTTP error into a structured form.
 * Handles various error body formats: empty, plain text, JSON, arbitrary bytes.
 * @param {number|null} status - HTTP status code
 * @param {string|null} bodyText - Decoded error body string
 * @param {string} message - Error message from the transport layer
 * @returns {NormalizedError}
 */
function normalizeError(status, bodyText, message) {
    const body = (bodyText || '').substring(0, MAX_ERROR_BODY_CHARS);
    // REVIEW-2026-11-07 (low): test the STORED (truncated) form, not the full
    // decoded text — mirrors _decode_body/normalize_http_error in
    // rca_core/error_utils.py. A message that embeds the full body also
    // embeds its truncated prefix, so this covers both cases.
    const messageCarriesBody = !body || message.indexOf(body) >= 0;
    // REVIEW-2026-09-20 #110: parse exactly what is stored (see
    // extractErrorCode()).
    const errorCode = extractErrorCode(body);
    const retryHint = getRetryHint(status, errorCode);

    return {
        status,
        body,
        message,
        messageCarriesBody,
        errorCode,
        retryHint,
    };
}

/**
 * Try to extract a machine-readable error code from the body.
 *
 * REVIEW-2026-09-20 #110 (mirror of rca_core/error_utils.py
 * _extract_error_code): callers must hand in the body AFTER the
 * MAX_ERROR_BODY_CHARS trim (normalizeError() does that), so the ``[CODE]``
 * prefix the user sees can only ever come from text that is actually
 * displayed — and a huge hostile body costs one bounded parse instead of a
 * second unbounded one. A body truncated mid-JSON simply yields no code,
 * like any other non-JSON error body.
 * @param {string|null} body - Already-truncated error body text
 * @returns {string|null}
 */
function extractErrorCode(body) {
    if (!body) return null;
    try {
        const data = JSON.parse(body);
        if (typeof data === 'object' && data !== null) {
            // Common error code locations
            for (const key of ['error_code', 'code', 'type', 'error.type']) {
                if (key in data) return String(data[key]);
            }
            // Nested error object
            const error = data.error;
            if (typeof error === 'object' && error !== null) {
                for (const key of ['error_code', 'code', 'type']) {
                    if (key in error) return String(error[key]);
                }
            }
        }
    } catch (_e) {
        // Not JSON, ignore
    }
    return null;
}

/**
 * Determine if and how to retry based on status and error code.
 * @param {number|null} status
 * @param {string|null} errorCode
 * @returns {string|null}
 */
function getRetryHint(status, errorCode) {
    if (status === null) return 'retry';  // Network error
    if (status === 429) return 'retry_with_delay';
    if (status === 401 || status === 403 || status === 400) return 'do_not_retry';
    if (status >= 500 && status < 600) return 'retry';
    if (RETRYABLE_STATUS.has(status)) return 'retry';
    return null;
}

/**
 * Check if an error should be retried.
 * @param {NormalizedError} error
 * @returns {boolean}
 */
function isRetryable(error) {
    if (error.status === null) return true;
    return RETRYABLE_STATUS.has(error.status);
}

/**
 * Get the best message for user display.
 * @param {NormalizedError} error
 * @returns {string}
 */
function getDisplayMessage(error) {
    let base;
    if (error.messageCarriesBody || !error.body) {
        base = error.message;
    } else {
        const prefix = error.status !== null ? `HTTP ${error.status}` : 'Network error';
        base = `${prefix}: ${error.body}`;
    }
    if (error.errorCode) {
        base = `[${error.errorCode}] ${base}`;
    }
    return base;
}

/**
 * Format a normalized error for user display.
 * @param {NormalizedError} error
 * @param {string|null} [prefix]
 * @returns {string}
 */
function formatError(error, prefix) {
    const msg = getDisplayMessage(error);
    if (prefix && error.status !== null) {
        return `${prefix} (${error.status}): ${msg}`;
    }
    if (error.status !== null) {
        return `HTTP ${error.status}: ${msg}`;
    }
    return msg;
}

// ---------------------------------------------------------------------------
// Retry with Exponential Backoff + Jitter
// ---------------------------------------------------------------------------

/**
 * Calculate retry delay with exponential backoff and jitter.
 * Respects Retry-After headers when present.
 * Mirrors DSH's provider-retry.js getRetryDelayMs() logic.
 *
 * @param {Object} options
 * @param {number|null} options.status - HTTP status code
 * @param {Headers|Object|null} options.headers - Response headers
 * @param {number} [options.attempt=0] - Current retry attempt (0-indexed)
 * @param {number} [options.initialDelay=0.8] - Initial backoff delay in seconds
 * @param {number} [options.backoffFactor=1.6] - Multiplicative backoff factor
 * @param {number} [options.maxDelay=60.0] - Maximum delay cap in seconds
 * @returns {number} Recommended delay in seconds before next retry
 */
function getRetryDelay({
    status,
    headers = null,
    attempt = 0,
    initialDelay = 0.8,
    backoffFactor = 1.6,
    maxDelay = 60.0,
}) {
    let delay = null;

    // 1. Check Retry-After-MS header (milliseconds)
    if (headers) {
        const retryAfterMs = headers.get && headers.get('retry-after-ms')
            ? headers.get('retry-after-ms')
            : headers['retry-after-ms'];
        if (retryAfterMs) {
            const parsed = parseFloat(retryAfterMs);
            if (!isNaN(parsed)) {
                delay = parsed / 1000.0;
            }
        }

        // 2. Check Retry-After header (seconds or HTTP date)
        if (delay === null) {
            const retryAfter = headers.get && headers.get('retry-after')
                ? headers.get('retry-after')
                : headers['retry-after'];
            if (retryAfter) {
                // Try parsing as float (seconds)
                const parsed = parseFloat(retryAfter);
                if (!isNaN(parsed)) {
                    delay = parsed;
                } else {
                    // Try parsing as HTTP date
                    try {
                        const date = new Date(retryAfter);
                        if (!isNaN(date.getTime())) {
                            const now = Date.now();
                            delay = Math.max(0, (date.getTime() - now) / 1000);
                        }
                    } catch (_e) {
                        // Not a valid date
                    }
                }
            }
        }
    }

    // 3. Fall back to exponential backoff with jitter
    if (delay === null) {
        const baseDelay = initialDelay * Math.pow(backoffFactor, attempt);
        // Add jitter: DSH pattern: multiply by (1 - random * 0.25)
        // This gives 0.75x to 1.0x of the base delay
        const jitter = 1.0 - Math.random() * 0.25;
        delay = baseDelay * jitter;
    }

    // REVIEW-2026-09-20 #108 (mirror of get_retry_delay()): cap first, then
    // lift to the 1 s floor. A ``Retry-After: 0`` (or an already-elapsed HTTP
    // date) is honoured as "no information" rather than as "hammer me again
    // immediately"; a caller that caps below the floor (``maxDelay = 0``, i.e.
    // the tests that must not sleep) still gets its ceiling because the floor
    // itself is clamped to it.
    delay = Math.min(delay, maxDelay);
    return Math.max(delay, Math.min(MIN_RETRY_DELAY_SECONDS, maxDelay));
}

/**
 * Sleep for a given number of milliseconds, respecting an AbortSignal.
 * @param {number} ms
 * @param {AbortSignal|null} signal
 * @returns {Promise<void>}
 */
function abortableSleep(ms, signal) {
    return new Promise((resolve, reject) => {
        if (signal?.aborted) {
            reject(new DOMException('Aborted', 'AbortError'));
            return;
        }
        const onAbort = () => {
            clearTimeout(timeout);
            reject(new DOMException('Aborted', 'AbortError'));
        };
        const timeout = setTimeout(() => {
            signal?.removeEventListener('abort', onAbort);
            resolve();
        }, Math.max(0, ms));
        signal?.addEventListener('abort', onAbort, { once: true });
    });
}

/**
 * Execute a function with exponential backoff retry.
 *
 * @param {Function} func - Async function that returns a result or throws
 * @param {Object} options
 * @param {number} [options.maxRetries=3]
 * @param {number} [options.initialDelay=0.8]
 * @param {number} [options.backoffFactor=1.6]
 * @param {number} [options.maxDelay=60.0]
 * @param {Function|null} [options.retryable=null] - Predicate: (result) => boolean
 * @param {Function|null} [options.onRetry=null] - Callback: (attempt, delay, error)
 * @param {AbortSignal|null} [options.signal=null] - AbortSignal to cancel retries
 * @returns {Promise<*>} The function's return value
 *
 * REVIEW-2026-09-20 #109 (mirror of retry_with_backoff(), and of the
 * `if attempt == retries - 1: return` guard in
 * rca_core/llm.py:call_llm_api_with_retry): the loop must NOT sleep once more
 * after the LAST attempt — previously the "result deemed retryable" path fell
 * through to the delay/sleep block even when no retry could follow, so a
 * 3-retry loop slept 4 times and ``on_retry`` reported a retry that never
 * happened. Both failure paths now ``break`` before the delay is computed, so
 * onRetry() and the sleep run exactly once per ACTUAL retry.
 *
 * The delay itself comes from getRetryDelay(), i.e. it carries the 1 s floor
 * (#108) — matching the two Python NETWORK retry loops (extractor.py
 * `_call_llm` and llm.py `call_llm_api_with_retry`, both of which go through
 * error_utils.get_retry_delay) — which is what this browser path mirrors.
 */
async function retryWithBackoff(func, {
    maxRetries = 3,
    initialDelay = 0.8,
    backoffFactor = 1.6,
    maxDelay = 60.0,
    retryable = null,
    onRetry = null,
    signal = null,
} = {}) {
    let lastError = null;
    let lastResult = null;

    for (let attempt = 0; attempt <= maxRetries; attempt++) {
        try {
            const result = await func();
            // Sprint B (REVIEW-2026-09-04): implement the documented
            // predicate semantics (mirrors the parallel fix to Python
            // rca_core/error_utils.retry_with_backoff). Previously BOTH
            // branches returned the result immediately, so the `retryable`
            // predicate was dead code:
            //   - retryable === null            -> accept the result;
            //   - retryable(result) === false   -> result is acceptable,
            //     return it immediately, no further retries;
            //   - retryable(result) === true    -> result warrants retry;
            //     keep retrying until retries are exhausted, then fall
            //     through and return the LAST result (no throw).
            if (retryable === null || !retryable(result)) {
                return result;
            }
            lastResult = result;
            if (attempt >= maxRetries) {
                break;
            }
        } catch (error) {
            lastError = error;
            if (attempt >= maxRetries) {
                break;
            }
        }

        const delay = getRetryDelay({
            status: lastError?.status || null,
            // FE-FIX-2026-09-22 (item 5 companion): transports that throw
            // with `err.headers` (js/minimax.js direct mode now surfaces
            // 429/408 for retry) get their Retry-After / Retry-After-Ms
            // honoured here — previously the loop only forwarded `status`,
            // so getRetryDelay's whole header branch was unreachable from
            // every caller of retryWithBackoff. Errors without headers keep
            // the exponential-backoff fallback (headers: null).
            headers: lastError?.headers || null,
            attempt,
            initialDelay,
            backoffFactor,
            maxDelay,
        });

        if (onRetry) {
            onRetry(attempt + 1, delay, lastError);
        }

        if (signal?.aborted) {
            throw new DOMException('Aborted', 'AbortError');
        }

        await abortableSleep(delay * 1000, signal);
    }

    // All retries exhausted
    if (lastError !== null) {
        throw lastError;
    }
    // The retryable predicate kept refusing every result. Return the last
    // result observed (null when func never produced one), matching the
    // Python docstring: "Returns None if all retries failed without an
    // exception".
    return lastResult;
}

// ---------------------------------------------------------------------------
// HTTP Error Extraction from fetch() Response
// ---------------------------------------------------------------------------

/**
 * Extract error info from a failed fetch() Response.
 * @param {Response} response
 * @returns {Promise<{status: number, body: string}>}
 */
async function extractFetchError(response) {
    let body = '';
    try {
        const text = await response.text();
        // Try to parse as JSON first
        try {
            const json = JSON.parse(text);
            // Extract meaningful error message
            if (json.error?.message) {
                body = json.error.message;
            } else if (json.error?.type) {
                body = json.error.type;
            } else if (json.message) {
                body = json.message;
            } else if (json.error) {
                body = typeof json.error === 'string' ? json.error : JSON.stringify(json.error);
            } else {
                body = text.substring(0, MAX_ERROR_BODY_CHARS);
            }
        } catch (_e) {
            // Not JSON, use raw text
            body = text.substring(0, MAX_ERROR_BODY_CHARS);
        }
    } catch (_e) {
        // Couldn't read body
        body = '';
    }
    return {
        status: response.status,
        body,
    };
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

// Export for use in other modules
window.RCAErrorUtils = {
    normalizeError,
    formatError,
    getDisplayMessage,
    isRetryable,
    getRetryHint,
    getRetryDelay,
    retryWithBackoff,
    abortableSleep,
    extractFetchError,
    // REVIEW-2026-09-20 #110 / #108: exported so the parity tests can pin the
    // truncated-body contract and the delay floor directly.
    extractErrorCode,
    RETRYABLE_STATUS,
    MAX_ERROR_BODY_CHARS,
    MIN_RETRY_DELAY_SECONDS,
};
